[Swift] Add a setting to control the Swift module loading mode

nathawes opened the pull request:
This adds a new setting that controls the Swift module loading mode, equivalent to the SWIFT_FORCE_MODULE_LOADING environment variable supported by the Swift compiler:

- `settings set symbols.swift-module-loading-mode only-parseable`
Only load swift modules via .swiftinterface files
- `settings set symbols.swift-module-loading-mode only-serialized`
Only load swift modules via .swiftmodule files
- `settings set symbols.swift-module-loading-mode prefer-parseable`
Try to load via a .swiftinterface, but fall back to the .swiftmodule if that fails (e.g. if it doesn't exist)
- `settings set symbols.swift-module-loading-mode prefer-serialized`
Try to load via a .swiftmodule, but fall back to the .swiftinterface if that fails (e.g. if it doesn't exist). This is the default.

This patch also updates lldb's `--repl` mode to respect the `-O` option so that it's possible to change this setting before the Swift module loader is initialized when using the REPL.

Commit:
[Swift] Support changing the Swift module loading mode via a new setting: symbols.swift-module-loading-mode

This setting controls the Swift module loading mode that gets passed to the
Swift module loader. Valid values are:

- only-parseable: only loads swift modules via .swiftinterface files
- only-serialized: only loads swift modules via .swiftmodule files
- prefer-parseable: tries to load via a .swiftinterface, but falls back to the
  .swiftmodule if it doesn't exist
- prefer-serialized: tries to load via a .swiftmodule, but falls back to the
  .swiftinterface if it exists or can't be loaded. This is the default.

nathawes commented:
@swift-ci test

nathawes commented:
@swift-ci test

nathawes commented:
@swift-ci test

nathawes commented:
@swift-ci test

jrose-apple commented on source/Symbol/SwiftASTContext.cpp:
```diff
@@ -2992,20 +2993,44 @@ swift::ASTContext *SwiftASTContext::GetASTContext() {
-      (swift::ParseableInterfaceModuleLoader *)parseable_module_loader_ap.get();
-      m_ast_context_ap->addModuleLoader(std::move(parseable_module_loader_ap));
+    auto MLM = swift::ModuleLoadingMode::PreferSerialized;
+    if (auto forceModuleLoadingMode =
```
Nitpick: this should use the LLDB variable naming style, even though it's mostly copied from Swift. (Alternately, we could wrap this up in a helper function that LLDB can call.)

jrose-apple commented on lit/SwiftREPL/SwiftInterfaceForceModuleLoadMode.test:
```diff
@@ -0,0 +1,65 @@
+// 1. Neither .swiftinterface or .swiftmodule present
+//
+// RUN: mv %t/lib %t/lib-backup
+// RUN: env SWIFT_FORCE_MODULE_LOADING=prefer-parseable %lldb --repl="-I%t/lib -L%t/lib -lAA -enable-parseable-module-interface" < %s 2>&1 | FileCheck %s -check-prefix=NOT-LOADED
```
I don't think you need the `-enable-parseable-module-interface` here; the environment variable implies it for the compiler and we're just always enabling it for LLDB.

nathawes commented on tools/driver/Driver.cpp:
```diff
@@ -662,100 +662,101 @@ int Driver::MainLoop() {
   bool handle_events = true;
   bool spawn_thread = false;
 
-  if (m_option_data.m_repl) {
```
The change here is just always taking the else branch, and instead checking if we're in REPL mode just before where we were previously calling m_debugger.RunCommandInterpreter(...) and calling m_debugger.RunREPL(...) instead. The rest is due to indentation change.

This was needed for --repl mode to still run the commands passed via -O (and written to commands_stream in the code above this line)

nathawes commented:
@swift-ci test

nathawes commented:
@swift-ci test

nathawes commented:
@swift-ci test

jrose-apple commented on source/Core/ModuleList.cpp:
```diff
@@ -85,12 +100,17 @@ static constexpr PropertyDefinition g_properties[] = {
+    {"swift-module-loading-mode", OptionValue::eTypeEnum, false,
+     eSwiftModuleLoadingModePreferSerialized, nullptr,
+     OptionEnumValues(g_swift_module_loading_mode_enums),
+     "The module loading mode to use when loading modules for Swift."}};
```
Is there a way to mark that this is only meant for testing?

nathawes replied:
I'm not sure, sorry. @adrian-prantl do you know if there's a way?

jrose-apple commented on tools/driver/Driver.cpp:
```diff
@@ -612,47 +612,51 @@ int Driver::MainLoop() {
-    commands_stream.Printf("\n");
+  // If we're not in --repl mode, add the commands to process the file
+  // arguments, and the commands specified to run afterwards.
+  if (!m_option_data.m_repl) {
```
Definitely want to make sure an LLDB person signs off on this.

adrian-prantl replied:
@jimingham I wanted to get your opinon on this, too. Currently`lldb --repl` always ignores `-O` options, but for testing they would be really useful.

Nathan made sure that the Swift driver always passes `-x` to `lldb --repl` to preserve the old behavior for users of `swift` but if you call `lldb --repl` you can now also specify `-O` options.

nathawes replied:
I think there was some confusion here, sorry. I *was* going to do that, but then I realized the existing behavior with `--repl` (and from swift) was actually already processing the .lldbinit from the home directory (it does so a little earlier in this function, before any of the command_stream/repl code changed here) and that behavior is actually useful as it gives us a way to set the new `swift-module-loading-mode` setting when running the repl via swift. I think this PR will change the behavior from swift, but only for users who have lldbinit files in their project directories and whose home directory lldbinit file sets the `target.load-cwd-lldbinit` setting to true (the default is to ignore them warn that they exist).

jimingham commented:
I'm fine with the code for the change, that looks ok.  Thanks for the tests!

I don't think many people run the REPL from the lldb driver, most folks do it by running swift.

So making the run through the lldb command line a bit more wild west is fine too.

I wonder, however, if there is any reason to allow the driver in --repl mode to handle commands after creating the target?  You could for instance do "-o run" and we'd get really confused...  This doesn't seem necessary for your purposes, so it would be nice to outlaw it.  IIUC, you only want to add settings that modify how we ingest the target.  The only other use for passing commands to lldb on the command line is  to add breakpoints in code you were planning to exercise in the REPL.  But you can set breakpoints before a target is created, and they will get added to the REPL target when it is made.  So -O will do for that as well.

So a good compromise might be to only allow -O and -S options when passing --repl, and not -o and -s.  That would be simple to implement, just don't do the call to WriteCommandsForSource(eCommandPlacementAfterFile, commands_stream).  If we do that we should also check that m_option_data.m_after_file_commands is empty and warning that we're going to drop those commands if not.  And we should drop a note into the driver's --help output to this effect.

nathawes commented:
@swift-ci please test

nathawes commented:
@swift-ci please test

adrian-prantl committed:
Update the lldb driver to support the -O and -S options when passing --repl

At the moment when --repl is passed to lldb it silently ignores any
commands passed via the options below:

--one-line-before-file <command>
                     Tells the debugger to execute this one-line lldb command before any file provided on the command line has been loaded.
--one-line <command>
                     Tells the debugger to execute this one-line lldb command after any file provided on the command line has been loaded.
--source-before-file <file>
                     Tells the debugger to read in and execute the lldb commands in the given file, before any file has been loaded.
--source <file>
                     Tells the debugger to read in and execute the lldb commands in the given file, after any file has been loaded.
-O <value>           Alias for --one-line-before-file
-o <value>           Alias for --one-line
-S <value>           Alias for --source-before-file
-s <value>           Alias for --source

The -O and -S options are quite useful when writing tests for the REPL
though, e.g. to change settings prior to entering REPL mode. This
patch updates the driver to still respect the commands supplied via -O
and -S when passing --repl instead of silently ignoring them. As -s
and -o don't really make sense in REPL mode, commands supplied via
those options are still ignored, but the driver now emits a warning to
make that clear to the user.

Patch by Nathan Hawes!

Differential Revision: https://reviews.llvm.org/D59681

git-svn-id: https://llvm.org/svn/llvm-project/lldb/trunk@356911 91177308-0d34-0410-b5e6-96231b3b80d8

nathawes commented:
@swift-ci please test

nathawes commented:
Updated this to use the upstreamed commit with the driver changes (supporting -O and -S with --repl). Thanks for your help with that @adrian-prantl!

jrose-apple commented on lit/Driver/TestRepl.test:
```diff
@@ -0,0 +1,7 @@
+# RUN: echo ':quit' | %lldb -x --repl -O 'expr 42' -S %S/Inputs/Print2.in -o 'expr 999999' -s %s 2>&1 | FileCheck %s
```
Nitpick: you can use `-b` to get LLDB to exit after it's processed all the -O/-o/-S/-s options.

adrian-prantl commented:
@swift-ci test and merge
